### ISSUE TITLE
libnd4j: MKL-DNN batch normalization given incorrect output for strided input

### ISSUE THREAD

**Issue by AlexDBlack** [deeplearning4j/deeplearning4j#7261]
Permuted input arrays (non-standard strides for input) gives incorrect results for mkl-dnn batch norm.
Yet strides clearly are obtained from the source array:
https://github.com/deeplearning4j/deeplearning4j/blob/master/libnd4j/include/ops/declarable/generic/nn/batchnorm.cpp#L48-L66

```
    @Test
    public void testBatchNormStidedInput(){

        for(boolean useMklDnn : new boolean[]{false, true}) {
            Nd4jCpu.Environment.getInstance().setUseMKLDNN(useMklDnn);

            INDArray in1 = Nd4j.rand('c', new int[]{2, 3, 12, 12});
            INDArray m = Nd4j.rand('c', new int[]{3});
            INDArray v = Nd4j.rand('c', new int[]{3});
            INDArray g = Nd4j.rand('c', new int[]{3});
            INDArray b = Nd4j.rand('c', new int[]{3});

            INDArray out1 = in1.ulike();
            DynamicCustomOp op = DynamicCustomOp.builder("batchnorm_new")
                    .addInputs(in1, m, v, g, b)
                    .addOutputs(out1)
                    .addIntegerArguments(1, 1, 1)
                    .addFloatingPointArguments(1e-3)
                    .build();
            Nd4j.exec(op);

            INDArray out2 = in1.ulike();
            INDArray in2 = Nd4j.create(in1.size(1), in1.size(0), in1.size(2), in1.size(3)).permute(1, 0, 2, 3);
            in2.assign(in1);
            assertTrue(in1.equalShapes(in2));
            assertEquals(in1, in2);

            op.setInputArgument(0, in2);
            op.setOutputArgument(0, out2);

            Nd4j.exec(op);

            assertEquals("Use mkldnn: " + useMklDnn, out1, out2);
        }
    }
```

**Comment by AlexDBlack** [comment]
Think I found it: source strides are used for destination strides. They don't need to be the same in the general case.
https://github.com/deeplearning4j/deeplearning4j/blob/master/libnd4j/include/ops/declarable/generic/nn/batchnorm.cpp#L226-L227

**Comment by saudet** [comment]
Are you sure it's running with float and not double data type, for example?
